chain/store/snapshot.go

```diff
@@ -68,6 +101,300 @@ func (cs *ChainStore) Import(ctx context.Context, r io.Reader) (*types.TipSet, e
 	return root, nil
 }
 
+type walkTask struct {
+	c        cid.Cid
+	taskType taskType
+}
+
+type walkResult struct {
+	c cid.Cid
+}
+
+type walkSchedulerConfig struct {
+	numWorkers      int64
+	tail            *types.TipSet
+	includeMessages bool
+	includeReceipts bool
+	includeState    bool
+}
+
+func newWalkScheduler(ctx context.Context, store bstore.Blockstore, cfg *walkSchedulerConfig, rootTasks ...*walkTask) (*walkScheduler, context.Context) {
+	tailSet := cid.NewSet()
+	for i := range cfg.tail.Cids() {
+		tailSet.Add(cfg.tail.Cids()[i])
+	}
+	grp, ctx := errgroup.WithContext(ctx)
+	s := &walkScheduler{
+		store:      store,
+		numWorkers: cfg.numWorkers,
+		stack:      rootTasks,
+		in:         make(chan *walkTask, cfg.numWorkers*64),
+		out:        make(chan *walkTask, cfg.numWorkers*64),
+		grp:        grp,
+		tail:       tailSet,
+		cfg:        cfg,
+	}
+	s.taskWg.Add(len(rootTasks))
+	return s, ctx
+}
+
+type walkScheduler struct {
+	store bstore.Blockstore
+	// number of worker routine to spawn
+	numWorkers int64
+	// buffer holds tasks until they are processed
+	stack []*walkTask
+	// inbound and outbound tasks
+	in, out chan *walkTask
+	// tracks number of inflight tasks
+	taskWg sync.WaitGroup
+	// launches workers and collects errors if any occur
+	grp *errgroup.Group
+	// set of tasks seen
+	seen sync.Map
+
+	tail *cid.Set
+	cfg  *walkSchedulerConfig
+}
+
+func (s *walkScheduler) enqueueIfNew(task *walkTask) {
+	if task.c.Prefix().MhType == mh.IDENTITY {
+		//log.Infow("ignored", "cid", todo.c.String())
+		return
+	}
+	if task.c.Prefix().Codec != cid.Raw && task.c.Prefix().Codec != cid.DagCBOR {
+		//log.Infow("ignored", "cid", todo.c.String())
+		return
+	}
+	if _, ok := s.seen.Load(task.c); ok {
+		return
+	}
+	log.Debugw("enqueue", "type", task.taskType.String(), "cid", task.c.String())
+	s.taskWg.Add(1)
+	s.seen.Store(task.c, struct{}{})
+	s.in <- task
+}
+
+func (s *walkScheduler) startScheduler(ctx context.Context) {
+	s.grp.Go(func() error {
+		defer func() {
+			log.Infow("walkScheduler shutting down")
+			close(s.out)
+			// Because the workers may have exited early (due to the context being canceled).
+			for range s.out {
+				s.taskWg.Done()
+			}
+			// Because the workers may have enqueued additional tasks.
+			for range s.in {
+				s.taskWg.Done()
+			}
+			// now, the waitgroup should be at 0, and the goroutine that was _waiting_ on it should have exited.
+			log.Infow("walkScheduler stopped")
```

Comment:
need to re-approach synchronization here. Seems last info message never happens. Things hang on ctrl-c.